File: 2381-shifting-letters-ii/2381-shifting-letters-ii.cpp
class Solution {

public:
    string shiftingLetters(string s, vector<vector<int>>& shifts) {
        auto delta = vector<int>(s.size() + 1);

        for (const auto& shift : shifts) {
            auto l = shift[0], r = shift[1], step = shift[2] == 1 ? 1 : -1;
            delta[l] += step;
            delta[r + 1] -= step;
        }
        int prev = 0;
        for (int i = 0; i < s.size(); ++i) {
            delta[i] += prev;
            s[i] = this->shift(s[i], delta[i]);
            prev = delta[i];
        }
        return s;
    }

private:
    inline char shift(char ch, int step) {
        return ((ch - 'a' + (step % 26) + 26) % 26) + 'a';
    }
};

File: 2381-shifting-letters-ii/2381-shifting-letters-ii.py
class Solution:
    def shiftingLetters(self, s: str, shifts: List[List[int]]) -> str:
        delta = [0 for _ in s]

        def tochr(i: int):
            base = ord("a")
            return chr(((i - base) + 26) % 26 + base)

        for l, r, direction in shifts:
            step = 1 if direction == 1 else -1

            delta[l] += step
            if r + 1 < len(delta):
                delta[r + 1] -= step

        for i in range(1, len(delta)):
            delta[i] += delta[i - 1]

        return "".join(tochr(ord(ch) + delta[i]) for i, ch in enumerate(s))
